File: apps/backend/app/schemas/workflow.py
"""
Workflow Engine Pydantic schemas.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional
from pydantic import BaseModel, ConfigDict, Field


class NodeBase(BaseModel):
    name: str
    node_type: str
    configuration: Optional[dict[str, Any]] = Field(default_factory=dict)
    timeout: float = Field(60.0, ge=1.0)


class NodeResponse(NodeBase):
    model_config = ConfigDict(from_attributes=True)
    id: int


class EdgeBase(BaseModel):
    source_node_id: int
    target_node_id: int
    transition_condition: Optional[str] = None
    priority: int = 0


class EdgeResponse(EdgeBase):
    model_config = ConfigDict(from_attributes=True)
    id: int


class WorkflowBase(BaseModel):
    name: str = Field(..., max_length=255)
    description: Optional[str] = None
    version: str = "1.0.0"


class WorkflowCreate(WorkflowBase):
    nodes: list[NodeBase] = Field(default_factory=list)
    edges: list[EdgeBase] = Field(default_factory=list)


class WorkflowResponse(WorkflowBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    organization_id: int
    status: str
    enabled: bool
    created_by: int
    created_at: datetime
    updated_at: datetime


class ExecutionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    workflow_id: int
    status: str
    current_node_id: Optional[int] = None
    started_at: datetime
    completed_at: Optional[datetime] = None
    duration: Optional[float] = None
    execution_context: dict[str, Any]

File: apps/frontend/src/integration/types/workflow.ts
export interface Node {
  id?: number;
  name: string;
  node_type: string;
  configuration?: Record<string, any>;
  timeout: number;
}

export interface Edge {
  id?: number;
  source_node_id: number;
  target_node_id: number;
  transition_condition?: string | null;
  priority: number;
}

export interface Workflow {
  id: number;
  name: string;
  description?: string | null;
  version: string;
  organization_id: number;
  status: string;
  enabled: boolean;
  created_by: number;
  created_at: string;
  updated_at: string;
}

export interface WorkflowCreate {
  name: string;
  description?: string;
  version: string;
  nodes: Node[];
  edges: Edge[];
}

export interface WorkflowExecution {
  id: number;
  workflow_id: number;
  status: string;
  current_node_id?: number | null;
  started_at: string;
  completed_at?: string | null;
  duration?: number | null;
  execution_context: Record<string, any>;
}
